src/crimson/os/seastore/async_cleaner.h

```diff
@@ -1036,6 +1046,52 @@ class SegmentCleaner : public SegmentProvider, public AsyncCleaner {
 
   clean_space_ret clean_space() final;
 
+  alloc_result_t alloc_new_extent(
```

Comment:
I don't follow the intention to move the entire logic into the AsyncCleaner, and inherit AsyncCleaner to ExtentAllocator.

I assume the original logic should be mostly unchanged in the EPM as it is deciding the generation, and only RBMCleaner has the ExtentAllocator feature for synchronous allocations.

A proper way might be to extend `ExtentOolWriter` with a new interface:
```
paddr_t alloc_paddr(extent_len_t length)
```
So that after the generation is decided in the original `alloc_new_extent()`, a proper `ExtentOolWriter` can be selected to allocate the paddr. This method is paired with `alloc_write_ool_extents()` to make sure the two methods for extents in the same generation are consistent -- either do non-delay allocations and write, or do delayed allocation and allocate-write. In other words, `SegmentedOolWriter::alloc_paddr()` should return `make_record_relative_paddr(0)`.

Reply:
Added.